src/cpp/src/threaded_streamer.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <condition_variable>
+#include <queue>
+#include <thread>
+
+#include "openvino/genai/llm_pipeline.hpp"
+#include "openvino/genai/tokenizer.hpp"
+#include "text_callback_streamer.hpp"
+
+namespace ov {
+namespace genai {
+
+class ThreadedStreamerWrapper {
+public:
+    ThreadedStreamerWrapper(const StreamerVariant& streamer, Tokenizer& tokenizer) {
+        if (auto streamer_obj = std::get_if<std::monostate>(&streamer)) {
+            m_streamer_ptr = nullptr;
+        } else if (auto streamer_obj = std::get_if<std::shared_ptr<StreamerBase>>(&streamer)) {
+            m_streamer_ptr = *streamer_obj;
+        } else if (auto callback = std::get_if<std::function<bool(std::string)>>(&streamer)) {
+            m_streamer_ptr = std::make_shared<TextCallbackStreamer>(tokenizer, *callback);
+        }
+    }
+
+    void start() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        m_worker_thread = std::make_shared<std::thread>(&ThreadedStreamerWrapper::_worker, this);
+    }
+
+    void put(const std::vector<int64_t>& tokens) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(tokens);
+        m_cv.notify_one();
+    }
+
+    void put(const int64_t token) {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_queue.push(token);
+        m_cv.notify_one();
+    }
+
+    void end() {
+        if (!m_streamer_ptr) {
+            return;
+        }
+
+        {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            m_stopped = true;
+        }
+
+        m_cv.notify_one();
+
+        if (m_worker_thread && m_worker_thread->joinable()) {
+            m_worker_thread->join();
+        }
+
+        m_streamer_ptr->end();
+    }
+
+    bool is_dropped() {
```

Comment:
should this method also have a check if callback is available?

Reply:
Yes, no need for unnecessary lock, added callback check.